### DISCUSSION TITLE
defer event handling until suitable state is reached

### DISCUSSION THREAD

**Discussion by davidhennemann** [statelyai/xstate#1305] [Q&A] [answered]
Lets say my machine is currently loading some resources from an external dependency. During this loading a user input occures and creates an xstate event, which currently is dropped, because its handled in the `loaded` state, but not in the `loading` state.

```javascript
Machine({
  initial: "loading",
  states: {
    loading: {
       invoke: {
          // loading logic here
          onDone: "loaded"
       },
    },
    loaded:  {
         on: {
               "USER_INTERACTION": processInteraction() // this might modify context etc...
         }
    }
  }

```
It would be really cool to have some kind of defer logic in place. In this example the loading state could defer the processing of the `USER_INTERACTION` event until a state is reached, which can handle this particular event.

Similar to the event deferral in UML state machines. https://en.wikipedia.org/wiki/UML_state_machine#Event_deferral

**Comment by Andarist** [comment]
I think this could be implemented in the userland quite easily by creating a helper like `withDeferredEvents` which would queue deferred events and `send` them using `pure` within `exit` handler. The only problematic part is this:
> In the case of orthogonal regions where one orthogonal region defers an event and another consumes the event, the consumer takes precedence and the event is consumed and not deferred.

But maybe being able to read the current state of the machine within action and having access to the config would allow to implement this as well in userland. 

Problems with including core support for this are:
- it's not translatable to SCXML directly
- adds to the bundlesize

**Comment by davidkpiano** [comment, marked as answer]
Here is one such implementation that you can do:

```js

const defer = () =>
  assign({
    $queue: (ctx, e) => ctx.$queue.concat(e)
  });

const recall = () =>
  actions.pure(ctx => {
    return ctx.$queue
      .map(deferredEvent => {
        return send(deferredEvent);
      })
      .concat(assign({ $queue: [] }));
  });

// Edit your machine(s) here
const machine = Machine({
  id: "machine",
  initial: "inactive",
  context: {
    $queue: []
  },
  states: {
    inactive: {
      on: { TOGGLE: "active", LOGIN: { actions: defer() } }
    },
    active: {
      entry: recall(),
      on: { TOGGLE: "inactive", LOGIN: "success" }
    },
    success: {}
  }
});

// Edit your service(s) here
const service = interpret(machine).onTransition(state => {
  console.log(state.value, state.event.type);
});

service.start();

service.send("LOGIN");  // inactive LOGIN (deferred)
service.send("TOGGLE"); // active TOGGLE
                        // active LOGIN (recalled)
```

https://codesandbox.io/s/flamboyant-noyce-xywi0?file=/src/index.js

**Reply by davidhennemann** [reply]
Thank you! 👍

**Reply by davidkpiano** [reply]
Edit: forgot to add `.concat(assign({ $queue: [] }));` - don't want you to get into an infinite loop!

**Comment by davidkpiano** [comment]
Reviving this topic to think of ideas for built-in deferred events. Apparently there's a lot of prior art:

- https://docs.spring.io/spring-statemachine/docs/1.1.x/reference/html/sm-deferevents.html
- https://www.state-machine.com/doc/Pattern_DeferredEvent.pdf
- https://stackoverflow.com/questions/53886486/deferring-events-from-sub-state-machine
- https://valelab4.ucsf.edu/svn/3rdpartypublic/boost/libs/msm/doc/HTML/ch02s02.html
- https://blog.wick.technology/spring-state-machine-deferred-events/

**Reply by codingedgar** [reply]
idk it this would help but:

akka actors can also stash incoming messages if they cannot handle them in the current state
https://doc.akka.io/docs/akka/current/typed/stash.html

also f# mailboxprocessor can read the inbox or scan it to process messages out of order:

```fsharp
# from https://riptutorial.com/fsharp/example/29148/out-of-order-message-processing
type Message =
    | RegularOperation of string
    | PriorityOperation of string

let processor = MailboxProcessor<Message>.Start(fun inbox ->
    let rec innerLoop () = async {
        let! priorityData = inbox.TryScan(fun msg ->
            // If there is a PriorityOperation, retrieve its data.
            match msg with
            | PriorityOperation data -> Some data
            | _ -> None)

        match priorityData with
        | Some data ->
            // Process the data from PriorityOperation.
        | None ->
            // No PriorityOperation was in the queue at the time, so
            // let's fall back to processing all possible messages
            let! message = inbox.Receive()
            match message with
            | RegularOperation data ->
                // We have free time, let's process the RegularOperation.
            | PriorityOperation data ->
                // We did scan the queue, but it might have been empty
                // so it is possible that in the meantime a producer
                // posted a new message and it is a PriorityOperation.
        // And never forget to process next messages.
        innerLoop ()
    }
    innerLoop())
   
  ``` 

I know this might feel like more in the domain of actors and less in the domain of state machines, but as XState is somewhat an actor, I'm not sure where to draw the line, however, this is what I think when I think about an async state machine, something that queues events, reads them (in various manners) and acts accordingly. But when I think of a sync state machine mailboxes are completely irrelevant. Yet I rarely make sync state machines and feel async states machines lack more fine-tuning with how messages are processed, I almost never want to process messages immediately.

**Comment by liuhenry** [comment]
Sorry to add-on to an old thread, but I figured I'd add an example of my (typed) conversion of @davidkpiano's [implementation](https://github.com/statelyai/xstate/discussions/1305#discussioncomment-37644) for v5:

```typescript
type Event =
  | { type: 'LOADED' }
  | // more events ... ;

export const machine = setup({
  types: {} as {
    context: {
      eventQueue: Event[];
    };
    events: Event;
  },
}).createMachine({
  initial: 'loading',
  context: {
    eventQueue: [],
  },
  states: {
    loading: {
      on: {
        LOADED: {
          target: 'ready',
        },
        '*': {
          actions: assign({
            eventQueue: ({ context: { eventQueue }, event }) => [
              ...eventQueue,
              event,
            ],
          }),
        },
      },
    },
    ready: {
      entry: enqueueActions(({ context: { eventQueue }, enqueue }) => {
        for (const event of eventQueue) {
          enqueue.raise(event);
        }
        enqueue.assign({
          eventQueue: [],
        });
      }),
      on: {
        // normal events
      },
    },
  },
});
```

Not sure if there's an alternative approach with the new APIs but this seems to work 👍
